mongo/gridfs/gridfs_examples_test.go

```diff
@@ -0,0 +1,225 @@
+	fileID, err := bucket.UploadFromStream("filename", bytes.NewBuffer(fileContent), uploadOpts)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	fmt.Printf("new file created with ID %s", fileID)
+}
+
+func ExampleBucket_OpenDownloadStream() {
+	var bucket *gridfs.Bucket
+	var fileID primitive.ObjectID
+
+	downloadStream, err := bucket.OpenDownloadStream(fileID)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer func() {
+		if err := downloadStream.Close(); err != nil {
+			log.Fatal(err)
+		}
+	}()
+
+	// Use SetReadDeadline to force a timeout if the download does not succeed in 2 seconds.
+	if err = downloadStream.SetReadDeadline(time.Now().Add(2 * time.Second)); err != nil {
+		log.Fatal(err)
+	}
+
+	fileBuffer := bytes.NewBuffer(nil)
+	if _, err := io.Copy(fileBuffer, downloadStream); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func ExampleBucket_DownloadToStream() {
+	var bucket *gridfs.Bucket
+	var fileID primitive.ObjectID
+
+	// Use SetReadDeadline to force a timeout if the download does not succeed in 2 seconds.
+	if err := bucket.SetReadDeadline(time.Now().Add(2 * time.Second)); err != nil {
+		log.Fatal(err)
+	}
+	// To make sure the deadline only applies to this download, set the deadline to an empty time.Time{} after the
+	// operation is done.
+	defer func() {
+		if err := bucket.SetReadDeadline(time.Time{}); err != nil {
+			log.Fatal(err)
+		}
+	}()
+
+	fileBuffer := bytes.NewBuffer(nil)
+	if _, err := bucket.DownloadToStream(fileID, fileBuffer); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func ExampleBucket_Delete() {
+	var bucket *gridfs.Bucket
+	var fileID primitive.ObjectID
+
+	// Use SetWriteDeadline to force a timeout if the upload does not succeed in 2 seconds.
+	if err := bucket.SetWriteDeadline(time.Now().Add(2 * time.Second)); err != nil {
+		log.Fatal(err)
+	}
+	// To make sure the deadline only applies to this operation, set the deadline to an empty time.Time{} after the
+	// operation is done.
+	defer func() {
+		if err := bucket.SetWriteDeadline(time.Time{}); err != nil {
+			log.Fatal(err)
+		}
+	}()
+
+	if err := bucket.Delete(fileID); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func ExampleBucket_Find() {
+	var bucket *gridfs.Bucket
+
+	// Specify a filter to find all files with a length greater than 1000 bytes.
+	filter := bson.D{
+		{"length", bson.D{{"$gt", 1000}}},
+	}
+	// Use SetReadDeadline to force a timeout if the upload does not succeed in 2 seconds.
+	if err := bucket.SetReadDeadline(time.Now().Add(2 * time.Second)); err != nil {
+		log.Fatal(err)
+	}
+	// To make sure the deadline only applies to this operation, set the deadline to an empty time.Time{} after the
+	// operation is done.
+	defer func() {
+		if err := bucket.SetReadDeadline(time.Time{}); err != nil {
+			log.Fatal(err)
+		}
+	}()
+
+	cursor, err := bucket.Find(filter)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer func() {
+		if err := cursor.Close(context.TODO()); err != nil {
+			log.Fatal(err)
+		}
+	}()
+
+	type gridfsFile struct {
```

Comment:
Should we provide a struct like this in the driver itself?

Reply:
Maybe? Not sure if other drivers do.

Reply:
Java (and .NET iirc) do.  Java's is called `com.mongodb.client.gridfs.model.GridFSFile`.

But that would be for another ticket anyway.